Explicitly instantiating `batch<>` and/or `batch_bool<>` causes compilation failures

Hiya,

Have been experimenting with ways to reduce compile times in our work codebase, thought I'd try `extern template`-ing some of the `batch<>` and `batch_bool<>` instances we use so they're only compiled in one TU. Unfortunately both classes fail to compile for a mixed bag of reasons when explicitly instantiated.

Simple repro:

```cpp
#include <xsimd/xsimd.hpp>

namespace xsimd
{
	template class batch<float>;
}

int main()
{
	return 0;
}
```

For the example above, MSVC, GCC and Clang all fail with some version of this:
```
././xsimd/config/../types/../types/xsimd_batch.hpp:698:24: error: no matching function for call to 'mod'
        return *this = kernel::mod<A>(*this, other, A {});
                       ^~~~~~~~~~~~~~
main.cpp:5:17: note: in instantiation of member function 'xsimd::batch<float>::operator%=' requested here
        template class batch<float>;
```

If you change the instantiation to a `batch_bool<>` of your choice, you instead get this on MSVC:
```
xsimd\types\xsimd_batch.hpp(272,1): error C2678: binary '=': no operator found which takes a left-hand
operand of type 'const xsimd::batch_bool<float,xsimd::default_arch>' (or there is no acceptable conversion)
```

Clang and GCC the same issue, but issue a better diagnostic:
```
././xsimd/config/../types/../types/xsimd_batch.hpp:270:89: error: no viable overloaded '='
        batch_bool& operator&=(batch_bool const& other) const noexcept { return (*this) = (*this) & other; }
                                                                                ~~~~~~~ ^ ~~~~~~~~~~~~~~~
main.cpp:5:17: note: in instantiation of member function 'xsimd::batch_bool<float>::operator&=' requested here
        template class batch_bool<float>;
```

I tested the above on SSE2, AVX and AVX2, though the nature of the errors don't appear ISA-related.

Not sure why these issues are only coming up when the templates are explicitly instantiated, might be that there's some internal template dots that otherwise aren't explicitly connected and the compilers are instantiating those parts lazily.

--- a/include/xsimd/types/xsimd_batch.hpp
+++ b/include/xsimd/types/xsimd_batch.hpp
@@ -21,6 +21,59 @@
 
 namespace xsimd
 {
+    template <class T, class A = default_arch>
+    class batch;
+    namespace types
+    {
+        template <class T, class A>
+        struct integral_only_operators
+        {
+            inline batch<T, A>& operator%=(batch<T, A> const& other) noexcept;
+            inline batch<T, A>& operator>>=(int32_t other) noexcept;
+            inline batch<T, A>& operator>>=(batch<T, A> const& other) noexcept;
+            inline batch<T, A>& operator<<=(int32_t other) noexcept;
+            inline batch<T, A>& operator<<=(batch<T, A> const& other) noexcept;
+
+            /** Shorthand for xsimd::mod() */
+            friend batch<T, A> operator%(batch<T, A> const& self, batch<T, A> const& other) noexcept
+            {
+                return batch<T, A>(self) %= other;
+            }
+
+            /** Shorthand for xsimd::bitwise_rshift() */
+            friend batch<T, A> operator>>(batch<T, A> const& self, batch<T, A> const& other) noexcept
+            {
+                return batch<T, A>(self) >>= other;
+            }
+
+            /** Shorthand for xsimd::bitwise_lshift() */
+            friend batch<T, A> operator<<(batch<T, A> const& self, batch<T, A> const& other) noexcept
+            {
+                return batch<T, A>(self) <<= other;
+            }
+
+            /** Shorthand for xsimd::bitwise_rshift() */
+            friend batch<T, A> operator>>(batch<T, A> const& self, int32_t other) noexcept
+            {
+                return batch<T, A>(self) >>= other;
+            }
+
+            /** Shorthand for xsimd::bitwise_lshift() */
+            friend batch<T, A> operator<<(batch<T, A> const& self, int32_t other) noexcept
+            {
+                return batch<T, A>(self) <<= other;
+            }
+        };
+        template <class A>
+        struct integral_only_operators<float, A>
+        {
+        };
+        template <class A>
+        struct integral_only_operators<double, A>
+        {
+        };
+
+    }
 
     /**
      * @brief batch of integer or floating point values.
@@ -31,8 +84,8 @@ namespace xsimd
      * @tparam T the type of the underlying values.
      * @tparam A the architecture this batch is tied too.
      **/
-    template <class T, class A = default_arch>
-    class batch : public types::simd_register<T, A>
+    template <class T, class A>
+    class batch : public types::simd_register<T, A>, public types::integral_only_operators<T, A>
     {
 
     public:
@@ -93,14 +146,9 @@ namespace xsimd
         inline batch& operator-=(batch const& other) noexcept;
         inline batch& operator*=(batch const& other) noexcept;
         inline batch& operator/=(batch const& other) noexcept;
-        inline batch& operator%=(batch const& other) noexcept;
         inline batch& operator&=(batch const& other) noexcept;
         inline batch& operator|=(batch const& other) noexcept;
         inline batch& operator^=(batch const& other) noexcept;
-        inline batch& operator>>=(int32_t other) noexcept;
-        inline batch& operator>>=(batch const& other) noexcept;
-        inline batch& operator<<=(int32_t other) noexcept;
-        inline batch& operator<<=(batch const& other) noexcept;
 
         // incr/decr operators
         inline batch& operator++() noexcept;
@@ -142,12 +190,6 @@ namespace xsimd
             return batch(self) /= other;
         }
 
-        /** Shorthand for xsimd::mod() */
-        friend batch operator%(batch const& self, batch const& other) noexcept
-        {
-            return batch(self) %= other;
-        }
-
         /** Shorthand for xsimd::bitwise_and() */
         friend batch operator&(batch const& self, batch const& other) noexcept
         {
@@ -164,30 +206,6 @@ namespace xsimd
         friend batch operator^(batch const& self, batch const& other) noexcept
         {
             return batch(self) ^= other;
-        }
-
-        /** Shorthand for xsimd::bitwise_rshift() */
-        friend batch operator>>(batch const& self, batch const& other) noexcept
-        {
-            return batch(self) >>= other;
-        }
-
-        /** Shorthand for xsimd::bitwise_lshift() */
-        friend batch operator<<(batch const& self, batch const& other) noexcept
-        {
-            return batch(self) <<= other;
-        }
-
-        /** Shorthand for xsimd::bitwise_rshift() */
-        friend batch operator>>(batch const& self, int32_t other) noexcept
-        {
-            return batch(self) >>= other;
-        }
-
-        /** Shorthand for xsimd::bitwise_lshift() */
-        friend batch operator<<(batch const& self, int32_t other) noexcept
-        {
-            return batch(self) <<= other;
         }
 
         /** Shorthand for xsimd::logical_and() */
@@ -726,10 +744,10 @@ namespace xsimd
     }
 
     template <class T, class A>
-    inline batch<T, A>& batch<T, A>::operator%=(batch<T, A> const& other) noexcept
+    inline batch<T, A>& types::integral_only_operators<T, A>::operator%=(batch<T, A> const& other) noexcept
     {
-        detail::static_check_supported_config<T, A>();
-        return *this = kernel::mod<A>(*this, other, A {});
+        ::xsimd::detail::static_check_supported_config<T, A>();
+        return *static_cast<batch<T, A>*>(this) = kernel::mod<A>(*static_cast<batch<T, A>*>(this), other, A {});
     }
 
     template <class T, class A>
@@ -754,31 +772,31 @@ namespace xsimd
     }
 
     template <class T, class A>
-    inline batch<T, A>& batch<T, A>::operator>>=(batch<T, A> const& other) noexcept
+    inline batch<T, A>& kernel::integral_only_operators<T, A>::operator>>=(batch<T, A> const& other) noexcept
     {
-        detail::static_check_supported_config<T, A>();
-        return *this = kernel::bitwise_rshift<A>(*this, other, A {});
+        ::xsimd::detail::static_check_supported_config<T, A>();
+        return *static_cast<batch<T, A>*>(this) = kernel::bitwise_rshift<A>(*static_cast<batch<T, A>*>(this), other, A {});
     }
 
     template <class T, class A>
-    inline batch<T, A>& batch<T, A>::operator<<=(batch<T, A> const& other) noexcept
+    inline batch<T, A>& kernel::integral_only_operators<T, A>::operator<<=(batch<T, A> const& other) noexcept
     {
-        detail::static_check_supported_config<T, A>();
-        return *this = kernel::bitwise_lshift<A>(*this, other, A {});
+        ::xsimd::detail::static_check_supported_config<T, A>();
+        return *static_cast<batch<T, A>*>(this) = kernel::bitwise_lshift<A>(*static_cast<batch<T, A>*>(this), other, A {});
     }
 
     template <class T, class A>
-    inline batch<T, A>& batch<T, A>::operator>>=(int32_t other) noexcept
+    inline batch<T, A>& kernel::integral_only_operators<T, A>::operator>>=(int32_t other) noexcept
     {
-        detail::static_check_supported_config<T, A>();
-        return *this = kernel::bitwise_rshift<A>(*this, other, A {});
+        ::xsimd::detail::static_check_supported_config<T, A>();
+        return *static_cast<batch<T, A>*>(this) = kernel::bitwise_rshift<A>(*static_cast<batch<T, A>*>(this), other, A {});
     }
 
     template <class T, class A>
-    inline batch<T, A>& batch<T, A>::operator<<=(int32_t other) noexcept
+    inline batch<T, A>& kernel::integral_only_operators<T, A>::operator<<=(int32_t other) noexcept
     {
-        detail::static_check_supported_config<T, A>();
-        return *this = kernel::bitwise_lshift<A>(*this, other, A {});
+        ::xsimd::detail::static_check_supported_config<T, A>();
+        return *static_cast<batch<T, A>*>(this) = kernel::bitwise_lshift<A>(*static_cast<batch<T, A>*>(this), other, A {});
     }
 
     /*****************************

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -142,6 +142,7 @@ set(XSIMD_TESTS
     test_complex_trigonometric.cpp
     test_conversion.cpp
     test_error_gamma.cpp
+    test_explicit_batch_instantiation.cpp
     test_exponential.cpp
     test_extract_pair.cpp
     test_fp_manipulation.cpp

--- a/test/test_explicit_batch_instantiation.cpp
+++ b/test/test_explicit_batch_instantiation.cpp
@@ -1,0 +1,32 @@
+/***************************************************************************
+ * Copyright (c) Johan Mabille, Sylvain Corlay, Wolf Vollprecht and         *
+ * Martin Renou                                                             *
+ * Copyright (c) QuantStack                                                 *
+ * Copyright (c) Serge Guelton                                              *
+ *                                                                          *
+ * Distributed under the terms of the BSD 3-Clause License.                 *
+ *                                                                          *
+ * The full license is in the file LICENSE, distributed with this software. *
+ ****************************************************************************/
+
+#include "xsimd/xsimd.hpp"
+#ifndef XSIMD_NO_SUPPORTED_ARCHITECTURE
+
+namespace xsimd
+{
+
+    template class batch<char>;
+    template class batch<unsigned char>;
+    template class batch<signed char>;
+    template class batch<unsigned short>;
+    template class batch<signed short>;
+    template class batch<unsigned int>;
+    template class batch<signed int>;
+    template class batch<unsigned long>;
+    template class batch<signed long>;
+    template class batch<float>;
+#if !XSIMD_WITH_NEON || XSIMD_WITH_NEON64
+    template class batch<double>;
+#endif
+}
+#endif

